src/v/net/conn_quota.cc

```diff
@@ -0,0 +1,440 @@
+            if (_cfg.max_connections()) {
+                vlog(
+                  rpc::rpclog.info,
+                  "Connection count limit updated to {}",
+                  _cfg.max_connections().value());
+                update_limit({}, total_home, _cfg.max_connections().value());
+            } else {
+                total_home = home_allowance{};
+                vlog(rpc::rpclog.info, "Connection count limit disabled");
+            }
+        } else {
+            if (!_cfg.max_connections()) {
+                // Reset, in anticipation of limit being re-enabled later
+                // (do not want to keep borrowed units around)
+                total_remote = remote_allowance{};
+            }
+        }
+    });
+
+    _cfg.max_connections_per_ip.watch([this]() {
+        if (!_cfg.max_connections_per_ip()) {
+            if (ss::this_shard_id() == 0) {
+                vlog(
+                  rpc::rpclog.info, "Connection count per-IP limit disabled");
+            }
+            ip_home.clear();
+            ip_remote.clear();
+        } else {
+            auto new_limit = _cfg.max_connections_per_ip().value();
+            if (ss::this_shard_id() == 0) {
+                vlog(
+                  rpc::rpclog.info,
+                  "Connection count per-IP limit updated to {}",
+                  new_limit);
+            }
+
+            for (auto& i : ip_home) {
+                update_limit(i.first, i.second, new_limit);
+            }
+        }
+    });
+}
+
+ss::future<> conn_quota::stop() { return _gate.close(); }
+
+/**
+ * Release a token, on close of a connection from client IP `addr`
+ *
+ * Always completes without yielding.  Usually shard local, but
+ * may spawn a background invoke_on future if the limits are close
+ * to being reached.
+ */
+void conn_quota::put(ss::net::inet_address addr) {
+    vlog(rpc::rpclog.trace, "put({})", addr);
+
+    // If enforcement was disabled since the token
+    // was issued, drop it on the floor.
+    if (_cfg.max_connections()) {
+        do_put({});
+    }
+
+    if (_cfg.max_connections_per_ip()) {
+        do_put(addr);
+    }
+}
+
+/**
+ * Attempt to acquire a token for an incoming connection from a particular
+ * client IP.  If none is available, an inert units object is returned.
+ *
+ * Very fast if limits are disabled.  If limits are enabled this is usually
+ * a shard-local operation but can be cross-shard if we are close to reaching
+ * limits.
+ */
+ss::future<conn_quota::units> conn_quota::get(ss::net::inet_address addr) {
+    vlog(rpc::rpclog.trace, "get({})", addr);
+
+    if (_cfg.max_connections()) {
+        if (!co_await do_get({})) {
+            co_return units();
+        };
+    } else {
+        vlog(rpc::rpclog.trace, "Global conn limit disabled");
+    }
+
+    if (_cfg.max_connections_per_ip()) {
+        if (!co_await do_get(addr)) {
+            // Release the unit we already took for total connection count
+            do_put({});
```

Comment:
does this mean that `_cfg.max_connections_per_ip()` being true implies `_cfg.max_connections()` being true? If so, then I guess that would be enforced some where otherwise it looks like the do_put/get would become imbalanced?

Reply:
No, but this was weird.  If max_connections was not set, but per_ip was, then we would effectively leak units into the total counter.  However, it wouldn't matter, because we weren't doing any enforcement of the total, and it gets reset in the event that max_connections is ever turned on.

Still, that's far too subtle, so I've changed this to make the `do_put` conditional on `max_connections`.